galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/spi/creds/FrameworkEncryptionService.java

```diff
@@ -0,0 +1,184 @@
+        }
+    }
+
+    public FrameworkEncryptionService(IFileSystem fileSystem, Environment environment) throws CredentialsException {
+        this(null, fileSystem, environment);
+    }
+
+    private EncryptionKeys parseEncryptionKeysFile() throws CredentialsException {
+        EncryptionKeys encryptionKeys = null;
+        String encryptionKeysLocation = environment.getenv(ENCRYPTION_KEYS_PATH_ENV);
+        if (encryptionKeysLocation != null) {
+            try {
+                String encryptionKeysYamlStr = fileSystem.readString(Paths.get(encryptionKeysLocation));
+                if (encryptionKeysYamlStr != null && !encryptionKeysYamlStr.isBlank()) {
+                    Yaml yamlParser = new Yaml(new Constructor(EncryptionKeys.class, new LoaderOptions()));
+                    encryptionKeys = yamlParser.load(encryptionKeysYamlStr);
+                }
+            } catch (Exception e) {
+                throw new CredentialsException("Failed to read encryption keys file", e);
+            }
+        }
+
+        return encryptionKeys;
+    }
+
+    private SecretKeySpec loadPrimaryEncryptionKey(EncryptionKeys encryptionKeysYaml) throws CredentialsException {
+        SecretKeySpec encryptionKey = null;
+        String encodedKeyStr = encryptionKeysYaml.getEncryptionKey();
+        if (encodedKeyStr != null) {
+            byte[] decodedPrimaryKeyBytes = Base64.getDecoder().decode(encodedKeyStr);
+            encryptionKey = new SecretKeySpec(decodedPrimaryKeyBytes, KEY_ALGORITHM);
+        }
+        return encryptionKey;
+    }
+
+    private List<SecretKeySpec> loadDecryptionKeys(EncryptionKeys encryptionKeysYaml) throws CredentialsException {
+        List<SecretKeySpec> decryptionKeys = new ArrayList<>();
+
+        // The encryption keys secret is mounted as a file of the form:
+        //   encryptionKey: <current-base64-encoded-key>
+        //   fallbackDecryptionKeys:
+        //   - <encoded-fallback-key-1>
+        //   - <encoded-fallback-key-2>
+        SecretKeySpec primaryEncryptionKey = loadPrimaryEncryptionKey(encryptionKeysYaml);
+        if (primaryEncryptionKey != null) {
+            decryptionKeys.add(primaryEncryptionKey);
+        }
+
+        for (String oldKey : encryptionKeysYaml.getFallbackDecryptionKeys()) {
+            byte[] oldKeyDecodedBytes = Base64.getDecoder().decode(oldKey);
+            SecretKeySpec key = new SecretKeySpec(oldKeyDecodedBytes, KEY_ALGORITHM);
+            decryptionKeys.add(key);
+        }
+        return decryptionKeys;
+    }
+
+    @Override
+    public String encrypt(String plainText) throws CredentialsException {
+        if (this.encryptionKey == null) {
+            throw new CredentialsException("Unable to encrypt the provided data. No encryption key has been set");
+        }
+
+        // Generate a random initialization vector (IV)
+        byte[] initVector = new byte[GCM_IV_BYTES_LENGTH];
+        SecureRandom secureRandom = new SecureRandom();
```

Comment:
If the randomizer isn't passed in separately, then unit tests are going to have a hard time creating reproducable results ? Seems that a Randomizer could be injected.